handles.go

```diff
@@ -106,18 +106,35 @@ func (b *BodyHandle) Size() int64 {
 }
 
 type BodyHandles struct {
-	lock    sync.RWMutex
-	handles []*BodyHandle
+	lock         sync.RWMutex
+	nextHandleID int
+	handles      map[int]*BodyHandle
+}
+
+func NewBodyHandles() *BodyHandles {
+	return &BodyHandles{handles: make(map[int]*BodyHandle)}
+}
+
+func (bhs *BodyHandles) getNextHandleID() int {
+	handleID := bhs.nextHandleID
+	bhs.nextHandleID += 1
+	return handleID
+}
+
+func (bhs *BodyHandles) addBodyHandle(bh *BodyHandle) (int, *BodyHandle) {
+	bhs.lock.Lock()
+	defer bhs.lock.Unlock()
+
+	id := bhs.getNextHandleID()
+	bhs.handles[id] = bh
+
+	return id, bh
 }
 
 func (bhs *BodyHandles) Get(id int) *BodyHandle {
 	bhs.lock.RLock()
 	defer bhs.lock.RUnlock()
 
-	if id >= len(bhs.handles) {
-		return nil
-	}
-
 	return bhs.handles[id]
```

Comment:
This returns nil if the key doesn't exist in the map, so it makes this simpler

Reply:
love it